### PR TITLE
Support Alternate Datetime Formats

### PR DESCRIPTION
**Pull request by GumpacG**
### Description
Supports using date formats obtained from OpenSearch. **Contains some changes from OpenSearchDataType rework (https://github.com/Bit-Quill/opensearch-project-sql/commit/8b2f65d1a9374de3b4ef526ab55cc5a8394b0231)**. Resolves some issues previously seen on the forum (please see _Issues Resolved_  section below). This also adds support to use custom formats in the index mapping. 

**TODO:** Follow up with support for date only, time only, and user-defined formats

**Based on PoC (Here: https://github.com/Bit-Quill/opensearch-project-sql/pull/169/files).**

Note: With these changes, issues may be introduced where data is read differently if a user's mapping defines a format. This is because the plugin previously used a hardcoded formatter to parse the data, while with these changes it will use the formatter defined in the mapping.

Supported formats include the following:

- basic_week_date_time
- strict_basic_week_date_time
- strict_week_date_time_no_millis
- basic_date_time_no_millis
- strict_week_date_time
- epoch_second
- basic_ordinal_date_time_no_millis
- ordinal_date_time
- strict_ordinal_date_time
- week_date_time
- week_date_time_no_millis
- strict_ordinal_date_time_no_millis
- ordinal_date_time_no_millis
- basic_date_time
- basic_week_date_time_no_millis
- strict_basic_week_date_time_no_millis

### Issues Resolved
https://github.com/opensearch-project/sql/issues/794
https://forum.opensearch.org/t/sql-select-fails-on-date-fields-format-epoch-second/11521

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by matthewryanwells — APPROVED** [review]

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java [hunk 1]
```diff
@@ -209,16 +236,14 @@ public String legacyTypeName() {
...
-    return copy;
+    return this.mappingType == null
+        ? new OpenSearchDataType(this.exprCoreType) : new OpenSearchDataType(this.mappingType);
   }
 
   /**
    * Flattens mapping tree into a single layer list of objects (pairs of name-types actually),
    * which don't have nested types.
-   * See {@link OpenSearchDataTypeTest#traverseAndFlatten() test} for example.
+   * See OpenSearchDataTypeTest#traverseAndFlatten() test for example.
```
Comment (forestmvey) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java:246]: Why remove the link?

**Comment by acarbonetto** [comment]
PR comment: 
> TODO: Follow up with support for date only, time only, and user-defined formats

TODO is stuff to do for this PR.  I would say instead: 
FOLLOW-UP PR:  support for date only, time only, and user-defined formats

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java [hunk 2]
```diff
@@ -97,56 +98,85 @@ public ExprType getExprType() {
...
+   */
+  public static Map<String, OpenSearchDataType> parseMapping(Map<String, Object> indexMapping) {
+    Map<String, OpenSearchDataType> result = new LinkedHashMap<>();
+
+    if (indexMapping == null) {
+      return result;
+    }
+
+    indexMapping.forEach((k, v) -> {
+      var innerMap = (Map<String, Object>)v;
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java:114]: if `v` is not a `(Map<String, Object>)` we should skip

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java [hunk 3]
```diff
@@ -97,56 +98,85 @@ public ExprType getExprType() {
...
-    ExprCoreType exprCoreType = mappingType.getExprCoreType();
-    if (exprCoreType == ExprCoreType.UNKNOWN) {
-      switch (mappingType) {
+  public static OpenSearchDataType of(MappingType mappingType, Map<String, Object> innerMap) {
+    OpenSearchDataType res = instances.getOrDefault(mappingType.toString(),
+        new OpenSearchDataType(mappingType)
+    );
+    switch (mappingType) {
+      case Object:
+        // TODO: use Object type once it has been added
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java:146]: Can we remove this TODO?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java [hunk 4]
```diff
@@ -97,56 +98,85 @@ public ExprType getExprType() {
...
-      }
+        Map<String, OpenSearchDataType> fields =
+            parseMapping((Map<String, Object>) innerMap.getOrDefault("fields", Map.of()));
+        return (!fields.isEmpty()) ? OpenSearchTextType.of(fields) : OpenSearchTextType.of();
+      case GeoPoint: return OpenSearchGeoPointType.of();
+      case Binary: return OpenSearchBinaryType.of();
+      case Ip: return OpenSearchIpType.of();
+      case Date:
+        return OpenSearchDateType.create(
+          (String) innerMap.getOrDefault("format", ""));
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java:166]: I would consider adding a comment that the `""` will use the default date format

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import lombok.EqualsAndHashCode;
+import org.opensearch.common.time.DateFormatter;
+import org.opensearch.sql.data.type.ExprType;
+
+/**
+ * Date type with support for predefined and custom formats read from the index mapping.
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java:19]: Do you know how you want to support `date` vs `time` vs `datetime`?  Do we want to use THIS object to cover all of these different types with explicit `exprCoreType`s for each, OR do you want to create separate classes for each one?  
I think we should add a TODO in the constructors.  e.g. 
```
// TODO:  Update exprCoreType depending on the format
```

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  @EqualsAndHashCode.Exclude
+  String formatString;
+
+  private OpenSearchDateType() {
+    super(MappingType.Date);
+    this.formatString = "";
+  }
+
+  private OpenSearchDateType(String formatStringArg) {
+    super(MappingType.Date);
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDateType.java:37]: you might consider adding a TODO here to support date vs time vs datetime

**Comment by GumpacG** [comment]
Closing because https://github.com/Bit-Quill/opensearch-project-sql/pull/268 was created to fix rebasing conflicts.

**Pull request closed without merging**